clang/lib/Sema/SemaTemplate.cpp

```diff
@@ -2792,6 +2811,24 @@ void DeclareImplicitDeductionGuidesForTypeAlias(
   } else {
     assert(false && "unhandled RHS type of the alias");
   }
+  return {Template, AliasRhsTemplateArgs};
+}
+
+// Build deduction guides for a type alias template.
+void DeclareImplicitDeductionGuidesForTypeAlias(
+    Sema &SemaRef, TypeAliasTemplateDecl *AliasTemplate, SourceLocation Loc) {
+  if (AliasTemplate->isInvalidDecl())
+    return;
+  auto &Context = SemaRef.Context;
+  // FIXME: if there is an explicit deduction guide after the first use of the
```

Comment:
What needs to happen to fix this? This is a decent sized issue.  Is this covered under UB (of so, can we diagnose that?), or does it have well defined rules?

Reply:
What about a diagnostic on an explicit deduction guide used after first use?  If they are not covered/added to the candidate set, we should diagnose that it isn't valid (then, anything AFTER the error where we generate them 'wrong' is irrelevant anyway).

Reply:
Emitting a diagnostic is an option, but I think it's suboptimal, and it is not trivial to detect such cases. I prefer to fix the issue directly rather than implementing a diagnostic that would eventually be removed.

Do you consider this case to be critical for the clang trunk at the moment? I believe we should fix it before the next clang release, and we still have some time to do so.

Reply:
WOULD we remove said diagnostic?  It seems that:
`If there are some explicit deduction guides after the first usage, they are not covered and are not added to the overload candidate sets.`

means that adding an explicit deduction guide after first use effectively 'does nothing'.  

That said, I'm ok having us do it (or whatever we do next) in a followup patch.